drivers/gpio/gpio_shell.c

```diff
@@ -21,167 +21,168 @@
-			  size_t argc, char **argv)
+static int cmd_gpio_blink(const struct shell *sh, size_t argc, char **argv)
 {
-	const struct device *dev;
-	uint8_t index = 0U;
-	uint8_t value = 0U;
+	struct sh_gpio gpio;
 	size_t count = 0;
+	int value = 0;
 	char data;
 
-	if (isdigit((unsigned char)argv[args_indx.index][0]) != 0) {
-		index = (uint8_t)atoi(argv[args_indx.index]);
-	} else {
-		shell_error(sh, "Wrong parameters for blink");
-		return -EINVAL;
-	}
-	dev = device_get_binding(argv[args_indx.port]);
+	get_sh_gpio(sh, argv, -2, &gpio);
 
-	if (dev != NULL) {
-		index = (uint8_t)atoi(argv[2]);
-		shell_fprintf(sh, SHELL_NORMAL, "Blinking port %s index %d.", argv[1], index);
-		shell_fprintf(sh, SHELL_NORMAL, " Hit any key to exit");
+	shell_fprintf(sh, SHELL_NORMAL, "Blinking port %s pin %u.", argv[-2], gpio.pin);
+	shell_fprintf(sh, SHELL_NORMAL, " Hit any key to exit");
 
-		/* dummy read to clear any pending input */
-		(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
+	/* dummy read to clear any pending input */
+	(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
 
-		while (true) {
-			(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
-			if (count != 0) {
-				break;
-			}
-			gpio_pin_set(dev, index, value);
-			value = !value;
-			k_msleep(SLEEP_TIME_MS);
+	while (true) {
+		(void)sh->iface->api->read(sh->iface, &data, sizeof(data), &count);
+		if (count != 0) {
+			break;
 		}
-
-		shell_fprintf(sh, SHELL_NORMAL, "\n");
+		gpio_pin_set(gpio.dev, gpio.pin, value);
+		value = !value;
+		k_msleep(SLEEP_TIME_MS);
 	}
 
+	shell_fprintf(sh, SHELL_NORMAL, "\n");
+
 	return 0;
 }
 
 SHELL_STATIC_SUBCMD_SET_CREATE(sub_gpio,
-			       SHELL_CMD_ARG(conf, NULL, "Configure GPIO", cmd_gpio_conf, 4, 0),
-			       SHELL_CMD_ARG(get, NULL, "Get GPIO value", cmd_gpio_get, 3, 0),
-			       SHELL_CMD_ARG(set, NULL, "Set GPIO", cmd_gpio_set, 4, 0),
-			       SHELL_CMD_ARG(blink, NULL, "Blink GPIO", cmd_gpio_blink, 3, 0),
+			       SHELL_CMD_ARG(conf, &sub_conf, "Configure GPIO", NULL, 2, 0),
+			       SHELL_CMD_ARG(get, NULL, "Get GPIO level", cmd_gpio_get, 1, 0),
+			       SHELL_CMD_ARG(set, &sub_set, "Set GPIO level", NULL, 2, 0),
+			       SHELL_CMD_ARG(blink, NULL, "Blink GPIO", cmd_gpio_blink, 1, 0),
 			       SHELL_SUBCMD_SET_END /* Array terminated. */
 			       );
 
-SHELL_CMD_REGISTER(gpio, &sub_gpio, "GPIO commands", NULL);
+static const char * const pin_syntax[] = {
+	 "0",  "1",  "2",  "3",  "4",  "5",  "6",  "7",  "8",  "9",
+	"10", "11", "12", "13", "14", "15", "16", "17", "18", "19",
+	"20", "21", "22", "23", "24", "25", "26", "27", "28", "29",
+	"30", "31"
+};
+
+static void cmd_gpio_pin_get(size_t idx, struct shell_static_entry *entry)
+{
+	if (idx < ARRAY_SIZE(pin_syntax)) {
+		entry->syntax  = pin_syntax[idx];
+		entry->handler = NULL;
+		entry->subcmd = &sub_gpio;
+		entry->help = "Pin";
+	} else {
+		entry->syntax = NULL;
+		entry->subcmd = &sub_gpio;
+	}
+}
```

Comment:
Would it be possible to remove `pin_syntax` and generate the list of available pins on the fly based on `DT_PROP(DT_NODELABEL(dev->name), ngpios)` ?

```suggestion
static void cmd_gpio_pin_get(size_t idx, struct shell_static_entry *entry)
{
	struct sh_gpio gpio;
	uint8_t ngpios;

	/* NOTE: Not sure how to retrieve 'sh' */

	get_sh_gpio(sh, argv, ARGV_GPIO_PORT - ARGV_GPIO_PIN, &gpio);
	ngpios = DT_PROP_OR(DT_NODELABEL(gpio.dev->name), ngpios, 0);

	if (idx < ngpios) {
		entry->syntax  = pin_syntax[idx];
		entry->handler = NULL;
		entry->subcmd = &sub_gpio;
		entry->help = "Pin";
	} else {
		entry->syntax = NULL;
		entry->subcmd = &sub_gpio;
	}
}
```

Reply:
@marco85mars20 yes, similar to this I've been using the Devicetree to generate the pin argument code, I'll hopefully push soon.